### PR TITLE
[Data] Improve reproducibility for random APIs

### LINKED ISSUE #55275: [Data/Train] Improve reproducibility
### Description

It seems that `ray.train.torch.enable_reproducibility()` should only be used inside a trainer function, not outside. This is because it sets some global seed, which affects other parts of the code that uses `np.random` or similar.

For example, `ray.data.read_parquet(..., shuffle="files")` will implicitly use the same seed. This becomes a problem in the training as the data has the same exact ordering in multiple epochs.

### Use case

In an ideal world, the model training (and data ingestion) remains usable (not just a short dev run) when we fix the random seed.

### PR DESCRIPTION
**Pull request by wingkitlee0**
## Description

This PR introduces `RandomSeedConfig` to handle random number generation initialization in `random_*` APIs and allow extra options for reproducibility across pipeline re-executions (e.g., training epochs). The class is modeled after `FileShuffleConfig` which uses `DataContext._execution_idx` as an extra seed.

When `RandomSeedConfig(seed, reseed_after_execution=True)` is set, the random ops will be re-seeded after each execution. This means that for random_* APIs, the results will be different for a new call to `iter_*_batches` APIs (i.e., `DataIterator`). In terms of model training, this means the data will be shuffle differently for each epoch, even when a initial random seed is set.

### Changes
* the following APIs now accept a `RandomSeedConfig` object as `seed`:
  * `random_sample()`
  * `random_shuffle()`
  * `randomize_block_order()`

## Related issues
Closes #55275

### Implementation

* Numpy's `default_rng()` takes an integer or a tuple of them. So when `reseed_after_execution` is set, we use `(execution_idx, base_seed)` to initialize. This needs to happens somewhere lower-level in Ray Data that is re-run during multiple execution.
* The seed tuple remains "deterministic" as the `execution_idx` is incremented by 1 each time the pipeline gets rerun. This in principle allows the multiple-epoch model training to be reproducible (minus other usual things that affect reproducibility, e.g., ordering, cluster configs, etc)
* This PR consolidates the `seed` argument. So the original integer/None seed is wrapped inside `RandomSeedConfig` before passing into the op definitions. In principle, we could make a default value of `reseed_after_execution` in `DataContext` to control all seed handling at once.

### Example

When `reseed_after_execution` is set to True, re-execution of the pipeline (second line) will return different results.
```python
ds = ray.data.range(10)
ds.random_shuffle(seed=RandomSeedConfig(1234, reseed_after_execution=True)).take_all()
```
Setting `reseed_after_execution` to `False` (or use integer seed directly) will produce the same (random) result when the pipeline is executed multiple times.

### Notes
* For a materialized Ray Dataset (i.e., `materialize()` is called), there is no effect as the pipeline is not re-executed in this case.
* There are a few other APIs with `seed` argument that I didn't touch.

### CODE REVIEW

**Comment by wingkitlee0** [comment]
bugbot run

File: python/ray/data/_internal/planner/randomize_blocks.py [hunk 1]
```diff
@@ -34,10 +43,9 @@ def fn(
         if len(blocks_with_metadata) == 0:
             return refs, {op._name: []}
         else:
-            if op._seed is not None:
-                random.seed(op._seed)
+            rng = np.random.default_rng(seed)
             input_owned = all(b.owns_blocks for b in refs)
-            random.shuffle(blocks_with_metadata)
+            rng.shuffle(blocks_with_metadata)
```
[the remark this answers is not in the document]
Comment (wingkitlee0) [reply python/ray/data/_internal/planner/randomize_blocks.py:45]: PCG64 is generally considered to be better than Mersenne Twister. Also we do not test for exact ordering. So the move from builtin random to numpy random sounds fine.

**Comment by wingkitlee0** [comment]
unstale

**Comment by wingkitlee0** [comment]
unstale please

File: python/ray/data/_internal/random_config.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+def get_timestamp_seed() -> int:
+    return time.time_ns() % INT32_MAX
+
+
+def get_single_integer_random_seed(
+    seed_config: RandomSeedConfig,
+    data_context: DataContext,
+    *,
+    use_timestamp_as_default: bool,
```
Comment (iamjustinhsu) [inline python/ray/data/_internal/random_config.py]: what are your thoughts on adding `use_timestamp_as_default` to `RandomSeedConfig`, and then in `randomize_blocks()` or `random_shuffle()` we generate the `RandomSeedConfig` with `use_timestamp_as_default=False/True`

File: python/ray/data/_internal/random_config.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        """
+        if self.seed is None:
+            return None
+        elif self.reseed_after_execution:
+            return (data_context._execution_idx, self.seed)
+        else:
+            return (self.seed,)
+
+    @classmethod
+    def _normalize_seed_argument(
```
Comment (iamjustinhsu) [inline python/ray/data/_internal/random_config.py]: do u think it would be better to rename to something like `create`?

File: python/ray/data/_internal/random_config.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        cls,
+        seed: Optional[Union[int, "RandomSeedConfig"]],
```
Comment (iamjustinhsu) [inline python/ray/data/_internal/random_config.py]: Since moving to python3.10, i think we now support the `|` syntax. There is no style guide for this, but thoughts on making it `int | RandomSeedConfig | None`? I just find it a bit more readable

File: python/ray/data/_internal/random_config.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        default_reseed_after_execution: bool,
+    ) -> "RandomSeedConfig":
+        """Return a RandomSeedConfig object based on the seed argument in
+        Ray Data API.
+
+        Args:
+            seed: A seed argument accepted by the Ray Data API.
```
Comment (iamjustinhsu) [inline python/ray/data/_internal/random_config.py]: I think this should be a bit more descriptive

**Review by iamjustinhsu — COMMENTED** [review]
nice work

**Comment by wingkitlee0** [comment]
@iamjustinhsu thanks for the review. I will come back to address your comments (and conflicts) when I get a chance.

File: python/ray/data/_internal/random_config.py (hunk 5 above)
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: Updated the docstring

File: python/ray/data/_internal/random_config.py (hunk 4 above)
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: It turns out I can't use str with | here. So I fixed it by `from __future__ import annotations`

File: python/ray/data/_internal/random_config.py (hunk 3 above)
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: Updated.

File: python/ray/data/_internal/random_config.py (hunk 2 above)
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: Good point. I have cleared some logic around this parameter in this and other functions as well, which are only used for "backward compatibility".

File: python/ray/data/tests/test_random_api.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+    if method == "randomize_block_order":
+        return ds.randomize_block_order(seed=seed)
+    raise ValueError(f"Unknown method: {method}")
+
+
+def test_random_api_deterministic_seed(base_dataset):
+    """Test that integer seed produces deterministic results."""
+    seed = 42
+
+    for method in ["random_shuffle", "random_sample", "randomize_block_order"]:
```
Comment (iamjustinhsu) [inline python/ray/data/tests/test_random_api.py]: i think u can create a pytest fixture for this, and then u can also remove all the pytest.mark.parameterize below
```python
@pytest.fixture(params=[...])
def method(request):
    return request.param

File: python/ray/data/tests/test_random_api.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+
+    assert (
+        result1 != result2
+    ), f"Different seeds should produce different results for {method}"
+
+
+@pytest.mark.parametrize(
+    "method", ["random_shuffle", "random_sample", "randomize_block_order"]
+)
+def test_random_api_none_seed(base_dataset, method):
```
Comment (iamjustinhsu) [inline python/ray/data/tests/test_random_api.py:98]: can this test be combined with the 2 above:
```python
@pytest.mark.parametrize(
    "seeds", [(None, None), (42,42), ...]
)
....
seed1, seed2 = seeds
if seed1 != seed2: then results should be diff, otherwise they should be the same?
```
syntax may be wrong, so lmk if that makes sense

File: python/ray/data/tests/test_random_api.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+    """Test reseed_after_execution behavior across multiple epochs."""
+    seed_value = 42
+    seed_false = RandomSeedConfig(seed=seed_value, reseed_after_execution=False)
+    seed_true = RandomSeedConfig(seed=seed_value, reseed_after_execution=True)
+
+    # Test random_shuffle with reseed_after_execution=False
+    ds = base_dataset.random_shuffle(seed=seed_false)
+    epochs = collect_batches_from_epochs(ds, num_epochs=3, batch_size=5)
+    assert (
+        epochs[0] == epochs[1] == epochs[2]
```
Comment (iamjustinhsu) [inline python/ray/data/tests/test_random_api.py:129]: curious, does this always succeed, like we don't need them to be sorted before comparison?

File: python/ray/data/tests/test_random_api.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        ), f"Fixed seed should produce same results across epochs for {method}"
+
+    elif seed_param == "RandomSeedConfig_true":
+        # reseed_after_execution=True should produce different results per epoch
+        # iter_batches() may trigger re-execution with different execution_idx
+        # Check all epochs are pairwise different by converting to tuples and using set
+        epoch_tuples = {tuple(epoch) for epoch in epochs}
+        assert len(epoch_tuples) == len(
+            epochs
+        ), f"reseed_after_execution=True should produce different results per epoch for {method}"
```
Comment (iamjustinhsu) [inline python/ray/data/tests/test_random_api.py]: can you elaborate why here u are checking the lengths, but the other 2 if statements are check for difference?

**Review by iamjustinhsu — COMMENTED** [review]
comments about the tests, but i think the src code is fine

**Review by iamjustinhsu — APPROVED** [review]
lgtm, the CI is failing, remember to rebase from master
cc: @bveeramani for review

File: python/ray/data/tests/test_random_api.py (hunk 8 above)
Comment (wingkitlee0) [reply python/ray/data/tests/test_random_api.py:129]: Updated docstrings

File: python/ray/data/tests/test_random_api.py (hunk 7 above)
Comment (wingkitlee0) [reply python/ray/data/tests/test_random_api.py:98]: Addressed in test_random_api_seed_determinism

**Comment by wingkitlee0** [comment]
Hi @iamjustinhsu @richardliaw , let me know if there are more comments or if this can be merged. Thanks so much for reviewing!

FYI, there was another PR closed for this: https://github.com/ray-project/ray/pull/61426

File: python/ray/data/_internal/random_config.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    seed_tuple = seed_config.get_seed_tuple(data_context=data_context)
+
+    if seed_tuple is None:
+        # This is a legacy behavior for some random operations.
+        return get_timestamp_seed() if seed_config.use_timestamp_as_default else None
+    elif len(seed_tuple) == 1:
+        return seed_tuple[0]
+
+    # Modulo ensures the result is in valid NumPy seed range [0, 2**32 - 1].
+    return hash(seed_tuple) % (2**32)
```
[the remark this answers is not in the document]
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: Noted, but I can't reproduce this potential issue. I will take a look later..

[the remark this answers is not in the document]
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: It turns out the hash() does not randomize for integers, which is what we used here. So this is fine.

Comment (goutamvenkat-anyscale) [inline python/ray/data/_internal/random_config.py]: Can we make this a const?

File: python/ray/data/_internal/random_config.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    def __post_init__(self):
+        """Ensure that the seed is either None or an integer."""
+        if self.seed is not None and not isinstance(self.seed, int):
+            raise ValueError("Seed must be an integer or None.")
+
+    def get_seed_tuple(
+        self,
+        *,
+        data_context: DataContext,
+    ) -> tuple[int] | tuple[int, int] | None:
```
Comment (goutamvenkat-anyscale) [inline python/ray/data/_internal/random_config.py]: Can we use a dataclass instead?

Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/random_config.py]: Return type should be something like `int | ExecutionIdSeed`

File: python/ray/data/dataset.py [hunk 12]
```diff
@@ -1940,18 +1990,22 @@ def random_sample(
...
 
+        seed_config = RandomSeedConfig.create_seed_config(seed)
+
         from ray.data._internal.execution.interfaces.task_context import TaskContext
 
-        def random_sample(batch: DataBatch, seed: Optional[int]):
+        def random_sample(batch: DataBatch, seed_config: RandomSeedConfig):
             ctx = TaskContext.get_current()
 
+            seed_tuple = seed_config.get_seed_tuple(data_context=self.context)
```
[the remark this answers is not in the document]
Comment (wingkitlee0) [reply python/ray/data/dataset.py]: Good catch. Fixed by passing data context directly.

File: python/ray/data/_internal/random_config.py (hunk 10 above)
Comment (wingkitlee0) [reply python/ray/data/_internal/random_config.py]: Fixed

File: python/ray/data/_internal/random_config.py [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+            Defaults to True.
+        use_timestamp_as_default: When enabled, it supports a legacy behavior that relies on
+            the timestamp as the default seed. This parameter is only used when the base seed
+            is None. Defaults to False. See ``get_single_integer_random_seed`` for more details.
+
+    """  # noqa: E501
+
+    seed: Optional[int] = None
+    reseed_after_execution: bool = True
+    use_timestamp_as_default: bool = False
```
[the remark this answers is not in the document]
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/random_config.py:79]: +1

**Review by goutamvenkat-anyscale — APPROVED** [review]
